test/types/hooks.test-d.ts

```diff
@@ -198,3 +205,18 @@ server.addHook('onReady', async function () {
 server.addHook('onClose', async (instance) => {
   expectType<FastifyInstance>(instance)
 })
+
+// Use case to monitor any regression on issue #3620
+// ref.: https://github.com/fastify/fastify/issues/3620
+const customTypedHook: preHandlerAsyncHookHandler<
+RawServerBase,
+RawRequestDefaultExpression<RawServerBase>,
+RawReplyDefaultExpression<RawServerBase>,
+Record<string, unknown>
+> = async function (_request, _reply) {
+  expectType<FastifyInstance>(this)
```

Comment:
can you add the assertion for FastifyRequest and FastifyReply as well?

Reply:
Done.
The test use case is now specifically targeting the failure in `this` typing.
I have seen in issues that linked to the original PR that it affects hooks with a wide range typing too.